File: 0486-predict-the-winner/0486-predict-the-winner.cpp
class Solution {
public:
    int n;
    vector<int> nums;
    bool PredictTheWinner(vector<int>& nums) {
        if(nums.size() == 1) return true;
        n = nums.size();
        this->nums = nums;
        vector<vector<int>> dp(n - 1, vector<int>(n, -1));
        int sum = 0;

        for(auto val : nums) {
            sum += val;
        }
        
        solve(0, n - 1, dp);

        for(auto v1 : dp) {
            for(auto v2: v1) {
                cout << v2 << "  ";
            }
            cout << endl;
        }
        return dp[0][n - 1] >= 0;
        
    }
    int solve(int i, int j, vector<vector<int>> &dp) {
        if(i > j) return 0;

        if(i == j) return nums[i];
        
        if(dp[i][j] != -1) return dp[i][j];
        
        int takeHead = nums[i] - solve(i + 1, j, dp);

        int takeTail = nums[j] - solve(i, j - 1, dp);
            
        return dp[i][j] = max(takeHead, takeTail);            
    }
};

File: 0486-predict-the-winner/0486-predict-the-winner.py
class Solution:
    def __init__(self):
        self.dp = [[-1] * 23 for _ in range(23)]
    
    def solve(self,i: int, j: int, nums: List[int]) -> int:
        if(i > j):
            return 0
        if(i == j):
            return nums[i]
        if(self.dp[i][j] != -1): return self.dp[i][j]
        takeHead = nums[i] - self.solve(i + 1, j, nums)
        takeTail = nums[j] - self.solve(i, j - 1, nums)
        self.dp[i][j] = max(takeHead, takeTail)        
        return self.dp[i][j]
    def PredictTheWinner(self, nums: List[int]) -> bool:
        n = len(nums)
        dp = [[-1 for _ in range(n)] for _ in range(n)]
        
        return self.solve(0, n - 1, nums) >= 0
        